---
title: "TF Classification"
author: "Arend Vancraeynest"
date: "`r Sys.Date()`"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE,
                      warning=FALSE, 
                      message=FALSE,
                      tidy.opts=list(width.cutoff=60)
                      )
```

```{r here}
library(here)
```

## Overview

Here, you must describe the goal of the analyses documented in this documented.

## Proteome metadata
### Ready the data for R manipulation
```{r}
library(Biostrings)
proteomes <- list.files(here("data", "Proteomes"), full.names = TRUE)

seqs <- lapply(proteomes, readAAStringSet)

# A file containing the genome assembly versions and reference papers has been made manually 
extra_data <- read.table(here("data", "Genome_references.txt"), header = FALSE, sep = "\t")
```


### Generate a summary dataframe
```{r}
# Initiate the dataframe based on the species names contained within the file names 
# (genus abbreviated in 3 or 4 letters, species epithet in full, divided by a '.')
df <- data.frame(Species_name=gsub(".*/([a-z]\\.[a-zA-Z]+)_((cv|ssp)\\.[a-z0-9-]+)?.*", "\\1 \\2", proteomes, ignore.case = TRUE))
df$Species_name <- gsub("\\.", ". ", df$Species_name)

# Add the Genome Assembly version
df$Genome_assembly_version <- extra_data$V1

# Count the number of proteins and genes for each species and append to an initiated dataframe
check_isoforms <- function(seqlist = NULL) {
  combined_list <- lapply(seq_along(seqlist), function(x){
    all_seqs <- names(seqlist[[x]])
    proteins <- length(all_seqs)
    
    # Substring the unique gene names from the different types of annotation
    unique_seqs <- gsub("^.*gene=", "", all_seqs, ignore.case = TRUE)
    unique_seqs <- gsub("\\s.*$", "", unique_seqs)
    unique_seqs <- gsub("\\.t?[0-9]+(.p)?$", "", unique_seqs)
    
    # Only retain unique gene handles
    unique_seqs <- unique(unique_seqs)
    genes <- length(unique_seqs)

    return(c(proteins, genes))
  }
  )
  df$Coding_genes <<- sapply(combined_list, "[[", 2)
  df$Annotated_proteins <<- sapply(combined_list, "[[", 1)
}

check_isoforms(seqs)

# Add the reference papers
df$References <- extra_data$V2

# Tidy up the dataframe to make it more reader-friendly by replacing underscores by spaces in the headers
names(df) <- gsub("_", " ", names(df))
```


### Clean up the protein FASTA files
```{r}
library(tidyverse)

primary_transcripts <- function(seq_stringset = NULL){
  n <- names(seq_stringset)
  
  # For each transcript, find the encoding gene
  unique_seqs <- gsub("^.*gene=", "", n, ignore.case = TRUE)
  unique_seqs <- gsub("\\s.*$", "", unique_seqs)
  unique_seqs <- gsub("\\.t?[0-9]+(.p)?$", "", unique_seqs)
  
  # Make a dataframe with the transcript names, gene names and sequence lengths
  # group by the gene names and only keep the longest sequence (which should by the primary)
  df_clean <- data.frame(names = n, clean_names = unique_seqs, len = width(seq_stringset)) %>% 
    group_by(clean_names) %>% 
    filter(len == max(len))
  
  # In some cases transcripts are duplicated (different id, but same length & sequence) --> keep only the first
  df_clean <- df_clean[!duplicated(df_clean$clean_names), ]
  
  # Return the AAStringset datatype filtered by the remaining names
  return(seq_stringset[df_clean$names])
}


# If the count of genes and proteins is not equal (aka more transcripts per gene instead of just the primary):
# apply the primary_transcript function. If equal: just copy the object
clean_seqlist <- lapply(seq_along(seqs), function(x){
  if(df[x,]$'Coding genes' != df[x,]$'Annotated proteins'){
    primary_transcripts(seqs[[x]])
  } else {seqs[[x]]
  }
})
```


## Session info

This document was created under the following conditions:

```{r sessioninfo}
sessionInfo()
```